### PR TITLE
WIP: Enhancements for PAC4J SAML Single Logout on CAS [5.1.x]

### PR DESCRIPTION
**Pull request by jkacer**
This PR contains enhancements that will allow proper SAML Single Logout using PAC4J.
The main problems to be solved here are:
- Properly construct the SAML Logout Request with data from the SAML2 profile, when the profile is not remembered between login and logout due to a missing session
- The profile will be saved on login and read just before logout into the current session
- The session will be invalidated at the very end of the logout flow, not earlier
- The 'service' parameter for logout will be ignored if SLO is active, to prevent redirects to the service before the SLO can be done

Still work in progress - incomplete, I'll be adding more changes.

The topic is being discussed here: https://groups.google.com/forum/#!topic/pac4j-dev/I2VdDIYgVNE

### CODE REVIEW

File: support/cas-server-support-pac4j-core-clients/src/test/java/org/apereo/cas/support/pac4j/web/flow/IgnoreServiceRedirectUrlForSamlActionTests.java [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+    private static final String CLIENT_NAME_SAML_NO_SLO = "SamlClientNoSlo";
+    private static final String CLIENT_NAME_SAML_WITH_SLO = "SamlClientWithSlo";
+    private static final String CLIENT_NAME_NON_SAML = "AnotherClient";
+
+    private IgnoreServiceRedirectUrlForSamlAction actionUnderTest;   
+
+
+    @Test
+    public void samlClientWithoutSloShouldNotRemoveService() throws Exception {
```
[the remark this answers is not in the document]
Comment (jkacer) [reply support/cas-server-support-pac4j-core-clients/src/test/java/org/apereo/cas/support/pac4j/web/flow/IgnoreServiceRedirectUrlForSamlActionTests.java]: Assert inside a method called in the test.

File: support/cas-server-support-pac4j-core-clients/src/test/java/org/apereo/cas/support/pac4j/web/flow/IgnoreServiceRedirectUrlForSamlActionTests.java [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        testServiceRemovalForParticularProfile(samlProfile(CLIENT_NAME_SAML_NO_SLO), false);
+    }
+
+
+    @Test
+    public void samlClientWithSloShouldRemoveService() throws Exception {
```
[the remark this answers is not in the document]
Comment (jkacer) [reply support/cas-server-support-pac4j-core-clients/src/test/java/org/apereo/cas/support/pac4j/web/flow/IgnoreServiceRedirectUrlForSamlActionTests.java]: Assert inside a method called in the test.

File: support/cas-server-support-pac4j-core-clients/src/test/java/org/apereo/cas/support/pac4j/web/flow/IgnoreServiceRedirectUrlForSamlActionTests.java [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        testServiceRemovalForParticularProfile(samlProfile(CLIENT_NAME_SAML_WITH_SLO), true);
+    }
+
+
+    @Test
+    public void otherClientShouldNotRemoveService() throws Exception {
```
[the remark this answers is not in the document]
Comment (jkacer) [reply support/cas-server-support-pac4j-core-clients/src/test/java/org/apereo/cas/support/pac4j/web/flow/IgnoreServiceRedirectUrlForSamlActionTests.java]: Assert inside a method called in the test.

**Comment by jkacer** [comment]
There are some Codacy and CheckStyle violation, I'll fix them later. Some of them have been already fixed.

**Comment by mmoayyed** [comment]
This is excellent. However note that we cannot ship this behavior in a patch release as you have it. Please reposition your PR to target master. Changes targetted for 5.1 need to be much smaller.

**Comment by jkacer** [comment]
Hi Misagh, sure, I'll prepare it for master. This will take some time...

**Comment by jkacer** [comment]
Hi Misagh, GitHub does not allow me to change the source branch from my repository, so I am going to close this PR and open a new one instead. The new one will be targeted at master.

**Comment by jkacer** [comment]
And here is the new pull request that goes against master: #3063.

**Pull request closed without merging**